ezmlm-idx.1
.\" $Id$
.TH ezmlm-idx 1
.SH NAME
ezmlm-idx \- create index for mailing list archive
.SH SYNOPSIS
.B ezmlm-idx
[
.B \-dDF
][
.B \-f\I msg
]
.I dir
.SH DESCRIPTION
.B ezmlm-idx
reads all archived messages for the list in
.I dir
from
.IR dir\fB/archive/
and creates a message index. The index file format is identical to
that produced by
.B ezmlm-send(1)
(when
.IR dir\fB/indexed
is present).

.B ezmlm-idx
will ignore messages that do not have the owner execute bit set.

.B ezmlm-idx
will create the the index file under a different name and then
move it into place.

If 
.IR dir\fB/indexed
does not exist,
.B ezmlm-idx
will create it to enable the use of the newly created subject and author index.

.B ezmlm-idx
will remove reply-indicators and the prefix from the subject before
entry into the index, as described for
.BR ezmlm-send(1) .
.B ezmlm-idx
will decode rfc2047 encoded subject and author headers. When unfolding
split lines,
.B ezmlm-idx
will remove redundant escape sequences for the character
set specified in
.IR dir\fB/charset .

.B ezmlm-idx
calculates a subject hash based on the normalized subject header. Normalization
removes linear whitespace, case information, and message reply indicators.
The subject hash is used by
.B ezmlm-get(1)
for message threading.
.SH OPTIONS
.TP
.B \-d
Use the ``Date:'' header for date information, rather than the top ``Received:''
header. This is unreliable, but useful when messages have been transferred from
another type of archive by mailing them to ezmlm.
.B ezmlm-idx
converts dates with two-digit years ``xx'' to 19xx if xx is >= 70. Otherwise,
20xx is assumed.
.TP
.B \-D
(Default.)
Use the ``Received:'' header, not the ``Date:'' header. This is more reliable
since this is the header added by the receiving qmail. Also, this results
in always correctly formatted dates in the same time zone (usually UTC).
.TP
.B \-f\fI msg
Start indexing with the archive section containing message
.IR msg .
This is useful if you have removed part of the archive and do not want to
create (empty) directories and index files for these. It also decreases
.B ezmlm-idx
memory use.
.TP
.B \-F
(Default.)
Start indexing from message 1.
.TP
.B \-v
Display
.B ezmlm-idx
version info.
.TP
.B \-V
Display
.B ezmlm-idx
version info.
.SH "SEE ALSO"
ezmlm-archive(1),
ezmlm-get(1),
ezmlm-send(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for SQLite3 subscriber databases.
  Thanks to Mike Tedder.

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Added support for decoding sender addresses mangled with a BATV "btv1" tag.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

- Renamed to "setup" target to the more standard "install".

- Modified ezmlm-split to exit silently if there is not "split" file,
  and added it back into the "manager" file to properly handle subscribe
  and unsubscribe requests destined for sublists.

- Made the ezmlm-reject -h (obey headerreject) option the default.

- Fixed ezmlm-send to obey the "addtrailer" setting instead of looking
  for a "text/trailer" file.

- Added support for custom subject lines to ezmlm-get (for digests).

- Fixed handling of adding the trailer on old lists.

- Fixed adding the trailer to posts encoded with base64.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.


ezmlm-idx.c
/*$Id: ezmlm-idx.c,v 1.29 1999/10/29 02:49:14 lindberg Exp $*/
/*$Name: ezmlm-idx-040 $*/

#include <sys/types.h>
#include <sys/stat.h>
#include "stralloc.h"
#include "subfd.h"
#include "strerr.h"
#include "error.h"
#include "lock.h"
#include "slurp.h"
#include "open.h"
#include "getln.h"
#include "sgetopt.h"
#include "case.h"
#include "scan.h"
#include "str.h"
#include "fmt.h"
#include "readwrite.h"
#include "exit.h"
#include "substdio.h"
#include "idx.h"
#include "mime.h"
#include "errtxt.h"
#include "getconf.h"
#include "makehash.h"

#define FATAL "ezmlm-idx: fatal: "

char strnum[FMT_ULONG];
char hash[HASHLEN];

stralloc fnadir = {0};
stralloc fnif = {0};
stralloc fnifn = {0};
stralloc fnaf = {0};

stralloc line = {0};
stralloc lines = {0};
stralloc dummy = {0};

int fdindexn;
int fdlock;
int fd;
int flagdate = 0;	/* use 'Received:' header by default, =1 -> 'Date:' */

	/* for reading index and in ezmlm-idx for reading message */
static substdio ssin;
static char inbuf[1024];

substdio ssindex;
char indexbuf[1024];

struct stat st;

stralloc subject = {0};
stralloc author = {0};
stralloc authmail = {0};
stralloc received = {0};
stralloc prefix = {0};
stralloc charset = {0};

struct strerr index_err;

stralloc num = {0};

char buf0[256];
substdio ss0 = SUBSTDIO_FDBUF(read,0,buf0,sizeof(buf0));

void die_usage()
{
  strerr_die1x(100,"ezmlm-idx: usage: ezmlm-idx [-dDF] [-f msg] dir");

}

void die_nomem()
{
  strerr_die2x(100,FATAL,ERR_NOMEM);
}

int idx_get_trimsubject()

/* reads an open message from 'fd', extracts the subject (if any), and    */
/* returns the subject in 'sub', the author in 'author', and the received */
/* rfc822 date to 'received'. 'fatal' is a program-specific error string. */
/* returns: 0 - no reply no prefix */
/*          1 - reply no prefix */
/*          2 - prefix no reply */
/*          3 - reply & prefix */
/* No terminal '\n' in any of the strallocs! */
{
char *cp;
int foundsubject = 0;
int issubject = 0;
int isfrom = 0;
int foundreceived = 0;
int foundfrom = 0;
int match;
int r;
unsigned int pos,pos1;

  substdio_fdbuf(&ssin,read,fd,inbuf,sizeof(inbuf));
  for (;;) {
    if (getln(&ssin,&line,&match,'\n') == -1)
      strerr_die2x(111,FATAL,ERR_READ_INPUT);
    if (match) {
      if (line.len == 1)
        break;
      if (*line.s == ' ' || *line.s == '\t') {
				/* continuation */
        if (issubject) {
          if (!stralloc_cat(&subject,&line)) die_nomem();
        } else if (isfrom)
          if (!stralloc_cat(&author,&line)) die_nomem();
      } else {
        issubject = 0;
        isfrom = 0;
        if (!foundsubject && case_startb(line.s,line.len,"Subject:")) {
          if (!stralloc_copyb(&subject,line.s+8,line.len-8)) die_nomem();
          foundsubject = 1;
          issubject = 1;
        } else if (!foundfrom && case_startb(line.s,line.len,"From:")) {
          if (!stralloc_copyb(&author,line.s+5,line.len-5)) die_nomem();
          foundfrom = 1;
          isfrom = 1;
        } else if (!flagdate && !foundreceived &&
            case_startb(line.s,line.len,"Received:")) {
          pos = byte_chr(line.s,line.len,';');
          if (pos != line.len)
            if (!stralloc_copyb(&received,line.s+pos+2,line.len - pos - 3))
              die_nomem();
          foundreceived = 1;
        } else if (flagdate && !foundreceived &&
            case_startb(line.s,line.len,"Date:")) {
          if (line.len < 22) continue;				/* illegal */
          pos = 6 + byte_chr(line.s+6,line.len-6,',');
          if (pos == line.len)
            pos = 5;
          ++pos;
          while (line.s[pos] == ' ' || line.s[pos] == '\t') ++pos;	/* dd */
          pos1 = pos + 3;
          while (++pos1 < line.len && line.s[pos1] != ' ');		/* mo */
          ++pos1;
          if (!stralloc_copyb(&received,line.s+pos,pos1 - pos))
              die_nomem();					/* '01 Jun ' */
          if (pos1 + 2 < line.len) {
            if (line.s[pos1 + 2] == ' ') {			/* 2-digit */
              if (line.s[pos1] >= '7') {			/* >= 70 */
              if (!stralloc_cats(&received,"19")) die_nomem();
              } else if (!stralloc_cats(&received,"20")) die_nomem();
              pos = pos1 + 3;					/* 2 digit */
            } else
              pos = pos1 + 5;					/* 4 digit */
            if (pos < line.len) {
              pos += byte_chr(line.s+pos,line.len-pos,' ');	/* after time */
              if (pos < line.len) {
                ++pos;						/* zone */
                while (line.s[pos] != ' ' && line.s[pos] != '\n') ++pos;
              } else
                pos = line.len - 1;	/* no zone. Illegal; better than 0 */
              if (!stralloc_catb(&received,line.s+pos1,pos - pos1))
			die_nomem();
              foundreceived = 1;
              continue;
            }
          }
          received.len = 0;		/* bad format - scrap */
        }
      }
    } else
      break;
  }

  if (foundsubject) {
    concatHDR(subject.s,subject.len,&lines,FATAL);	/* make 1 line */
    decodeHDR(lines.s,lines.len,&line,charset.s,FATAL);	/* decode mime */
    r= unfoldHDR(line.s,line.len,&subject,charset.s,&prefix,1,FATAL);
						 /* trim mime */
  }
  else {
    r = 0;
    subject.len = 0;
  }
  return r;
}

int main(argc,argv)
int argc;
char **argv;
{
  char *dir,*cp;
  unsigned long msgnum = 0L;
  unsigned long msgmax;
  int opt,r;

  while ((opt = getopt(argc,argv,"dDf:FvV")) != opteof)
    switch (opt) {
      case 'd': flagdate = 1; break;
      case 'D': flagdate = 0; break;
      case 'f': if (optarg) (void) scan_ulong(optarg,&msgnum); break;
      case 'F': msgnum = 0L;
      case 'v':
      case 'V': strerr_die2x(0,"ezmlm-archive version: ",EZIDX_VERSION);
      default: die_usage();
  }
  dir = argv[optind];
  if (!dir) die_usage();

  if (chdir(dir) == -1)
    strerr_die4sys(100,FATAL,ERR_SWITCH,dir,": ");

  (void) umask(022);
  sig_pipeignore();
			/* obtain lock to write index files */
  fdlock = open_append("lock");
  if (fdlock == -1)
    strerr_die2sys(100,FATAL,ERR_OPEN_LOCK);
  if (lock_ex(fdlock) == -1)
    strerr_die2sys(100,FATAL,ERR_OBTAIN_LOCK);

  getconf_line(&charset,"charset",0,FATAL,dir);
  if (!stralloc_0(&charset)) die_nomem();

  getconf_line(&prefix,"prefix",0,FATAL,dir);
					/* support rfc2047-encoded prefix */
  decodeHDR(prefix.s,prefix.len,&line,charset.s,FATAL);
  unfoldHDR(line.s,line.len,&prefix,charset.s,&dummy,0,FATAL);
					/* need only decoded one */

			/* Get message number */
  switch(slurp("num",&num,32)) {
    case -1:
      strerr_die4sys(100,FATAL,ERR_READ,dir,"/num: ");
    case 0:
      strerr_die4x(100,FATAL,dir,"/num",ERR_NOEXIST);
  }
  if (!stralloc_0(&num)) die_nomem();
  scan_ulong(num.s,&msgmax);
  if (msgnum > msgmax) _exit(0);
  if (msgnum) {
    msgnum = (msgnum / 100) * 100 - 1;
  }
  while (++msgnum <= msgmax) {
    if (msgnum == 1 || !(msgnum % 100)) {
      if (!stralloc_copys(&fnadir,"archive/")) die_nomem();
      if (!stralloc_catb(&fnadir,strnum,fmt_ulong(strnum,msgnum / 100)))
	die_nomem();
      if (!stralloc_copy(&fnifn,&fnadir)) die_nomem();
      if (!stralloc_copy(&fnif,&fnadir)) die_nomem();
      if (!stralloc_cats(&fnif,"/index")) die_nomem();
      if (!stralloc_cats(&fnifn,"/indexn")) die_nomem();
      if (!stralloc_0(&fnadir)) die_nomem();
      if (!stralloc_0(&fnifn)) die_nomem();
      if (!stralloc_0(&fnif)) die_nomem();

			/* May not exist, so be nice and make it */
      if (mkdir(fnadir.s,0755) == -1)
	if (errno != error_exist)
	  strerr_die4sys(100,FATAL,ERR_CREATE,fnadir.s,": ");

			/* Open index */
      fdindexn = open_trunc(fnifn.s);
      if (fdindexn == -1)
        strerr_die4sys(100,FATAL,ERR_WRITE,fnifn.s,": ");

			/* set up buffers for index */
      substdio_fdbuf(&ssindex,write,fdindexn,indexbuf,sizeof(indexbuf));

			/* Get subject without the 'Subject: ' */
			/* make sure there is one */
    }

    if (!stralloc_copys(&fnaf,fnadir.s)) die_nomem();
    if (!stralloc_cats(&fnaf,"/")) die_nomem();
    if (!stralloc_catb(&fnaf,strnum,
       fmt_uint0(strnum,(unsigned int) (msgnum % 100),2))) die_nomem();
    if (!stralloc_0(&fnaf)) die_nomem();
    fd = open_read(fnaf.s);
    if (fd == -1) {
      if (errno != error_noent)
        strerr_die4sys(100,FATAL,ERR_READ,fnaf.s,": ");
    } else if (fstat(fd,&st) == -1 || (!(st.st_mode & 0100)))
        close(fd);
    else {
      int k;

      subject.len = 0;		/* clear in case they're missing in msg */
      author.len = 0;
      received.len = 0;
      r = idx_get_trimsubject();
      close(fd);
      if (!stralloc_copyb(&line,strnum,fmt_ulong(strnum,msgnum))) die_nomem();
      if (!stralloc_cats(&line,": ")) die_nomem();
      makehash(subject.s,subject.len,hash);
      if (!stralloc_catb(&line,hash,HASHLEN)) die_nomem();
      if (!stralloc_cats(&line," ")) die_nomem();
      if (r & 1)	/* reply */
	if (!stralloc_cats(&line,"Re: ")) die_nomem();
      if (!stralloc_cat(&line,&subject)) die_nomem();
      if (!stralloc_cats(&line,"\n\t")) die_nomem();
      if (!stralloc_cat(&line,&received)) die_nomem();
      if (!stralloc_cats(&line,";")) die_nomem();

      concatHDR(author.s,author.len,&lines,FATAL);
      mkauthhash(lines.s,lines.len,hash);
      if (!stralloc_catb(&line,hash,HASHLEN)) die_nomem();

      k = author_name(&cp,lines.s,lines.len);
      decodeHDR(cp,k,&author,charset.s,FATAL);

      (void) unfoldHDR(author.s,author.len,&lines,charset.s,&prefix,0,FATAL);

      if (!stralloc_cats(&line," ")) die_nomem();
      if (!stralloc_cat(&line,&lines)) die_nomem();
      if (!stralloc_cats(&line,"\n")) die_nomem();
      if (substdio_put(&ssindex,line.s,line.len) == -1)
          strerr_die4sys(100,FATAL,ERR_WRITE,fnifn.s, ": ");
    }

    if (!((msgnum + 1) % 100) ||
		(msgnum == msgmax)) {	/* last in this set */
      if (substdio_flush(&ssindex) == -1)
        strerr_die4sys(100,FATAL,ERR_FLUSH,fnifn.s, ": ");
      if (fsync(fdindexn) == -1)
        strerr_die4sys(100,FATAL,ERR_SYNC,fnifn.s, ": ");
      if (fchmod(fdindexn,MODE_ARCHIVE | 0700) == -1)
        strerr_die4sys(100,FATAL,ERR_WRITE,fnifn.s, ": ");
      if (close(fdindexn) == -1)
        strerr_die4sys(100,FATAL,ERR_CLOSE,fnifn.s,": ");
      if (rename(fnifn.s,fnif.s) == -1)
        strerr_die4x(111,FATAL,ERR_MOVE,fnifn.s,": ");
    }
  }
  fd = open_append("indexed");
  if (fd == -1)
    strerr_die4sys(100,FATAL,ERR_CREATE,dir,"/indexed: ");
  close(fd);
  close(fdlock);
  _exit(0);
}

